frame/assets/src/lib.rs

```diff
@@ -569,45 +603,89 @@ pub mod pallet {
 			Self::do_force_create(id, owner, is_sufficient, min_balance)
 		}
 
-		/// Destroy a class of fungible assets.
+		/// Start the process of destroying a class of fungible asset
+		/// start_destroy is the first in a series of extrinsics that should be called, to allow
+		/// destroying an asset.
 		///
 		/// The origin must conform to `ForceOrigin` or must be Signed and the sender must be the
 		/// owner of the asset `id`.
 		///
 		/// - `id`: The identifier of the asset to be destroyed. This must identify an existing
-		/// asset.
-		///
-		/// Emits `Destroyed` event when successful.
-		///
-		/// NOTE: It can be helpful to first freeze an asset before destroying it so that you
-		/// can provide accurate witness information and prevent users from manipulating state
-		/// in a way that can make it harder to destroy.
-		///
-		/// Weight: `O(c + p + a)` where:
-		/// - `c = (witness.accounts - witness.sufficients)`
-		/// - `s = witness.sufficients`
-		/// - `a = witness.approvals`
-		#[pallet::weight(T::WeightInfo::destroy(
-			witness.accounts.saturating_sub(witness.sufficients),
- 			witness.sufficients,
- 			witness.approvals,
- 		))]
-		pub fn destroy(
+		///   asset.
+		///
+		/// Assets must be freezed before calling start_destroy.
+		#[pallet::weight(T::WeightInfo::start_destroy())]
+		pub fn start_destroy(
 			origin: OriginFor<T>,
 			#[pallet::compact] id: T::AssetId,
-			witness: DestroyWitness,
-		) -> DispatchResultWithPostInfo {
+		) -> DispatchResult {
 			let maybe_check_owner = match T::ForceOrigin::try_origin(origin) {
 				Ok(_) => None,
 				Err(origin) => Some(ensure_signed(origin)?),
 			};
-			let details = Self::do_destroy(id, witness, maybe_check_owner)?;
-			Ok(Some(T::WeightInfo::destroy(
-				details.accounts.saturating_sub(details.sufficients),
-				details.sufficients,
-				details.approvals,
-			))
-			.into())
+			Self::do_start_destroy(id, maybe_check_owner)
+		}
+
+		/// Destroy all accounts associated with a given asset.
+		/// `destroy_accounts` should only be called after `start_destroy` has been called, and the
+		/// asset is in a `Destroying` state
+		///
+		/// Due to weight restrictions, this function may need to be called multiple
+		/// times to fully destroy all accounts. It will destroy `RemoveItemsLimit` accounts at a
+		/// time.
+		///
+		/// - `id`: The identifier of the asset to be destroyed. This must identify an existing
+		///   asset.
+		///
+		/// Each call Emits the `Event::DestroyedAccounts` event.
+		#[pallet::weight(T::WeightInfo::destroy_accounts(T::RemoveItemsLimit::get()))]
+		pub fn destroy_accounts(
+			origin: OriginFor<T>,
+			#[pallet::compact] id: T::AssetId,
```

Comment:
```suggestion
			origin: OriginFor<T>,
			#[pallet::compact] id: T::AssetId,
			max_items: u32,
```
Otherwise it is inflexible and could clog up the blocks.

Reply:
@ggwpez I've made a slight change, but the extrinsic still depends on the RemoveItemsLimit value declared on the pallet config.  That's basically how it gets the value of max_items, and not as an argument from the caller of the extrinsic

Reply:
@tonyalaribe by adding this extra param `max_items` we optimise the logic. If for example the RemoveItemsLimit will be set to u32::MAX, and we actually need to remove 10 records only, then we would reserve much more space than actually needed.
You would also need to add an extra check then like: `ensure!(max_items <= T::RemoveItemsLimit::get(), Error::<T, I>::SomeError);`